video/out/opengl/hwdec_drmprime_egl.c

```diff
@@ -0,0 +1,279 @@
+    EGLImageKHR images[MAX_NUM_PLANES];
+
+    EGLImageKHR (EGLAPIENTRY *CreateImageKHR)(EGLDisplay, EGLContext,
+                                              EGLenum, EGLClientBuffer,
+                                              const EGLint *);
+    EGLBoolean (EGLAPIENTRY *DestroyImageKHR)(EGLDisplay, EGLImageKHR);
+    void (EGLAPIENTRY *EGLImageTargetTexture2DOES)(GLenum, GLeglImageOES);
+};
+
+static void uninit(struct ra_hwdec *hw)
+{
+}
+
+static int get_egl_mp_image_colorspace_attr(struct mp_image *mpi)
+{
+    switch(mpi->params.color.space) {
+        case MP_CSP_BT_601 :
+            return EGL_ITU_REC601_EXT;
+
+        case MP_CSP_BT_709 :
+            return EGL_ITU_REC709_EXT;
+
+        case MP_CSP_BT_2020_C :
+        case MP_CSP_BT_2020_NC :
+            return EGL_ITU_REC2020_EXT;
+        default:
+            return EGL_ITU_REC601_EXT;
+    }
+}
+
+static int init(struct ra_hwdec *hw)
+{
+    if (!ra_is_gl(hw->ra) || !eglGetCurrentContext())
+        return -1;
+
+    const char *exts = eglQueryString(eglGetCurrentDisplay(), EGL_EXTENSIONS);
+    if (!exts)
+        return -1;
+
+    GL *gl = ra_gl_get(hw->ra);
+    if (!strstr(exts, "EXT_image_dma_buf_import") ||
+        !strstr(exts, "EGL_KHR_image_base") ||
+        !strstr(gl->extensions, "GL_OES_EGL_image"))
+    {
+        MP_ERR(hw, "EGL doesn't support the following extensions : "
+                   "EXT_image_dma_buf_import, EGL_KHR_image_base, "
+                   "GL_OES_EGL_image\n");
+        return -1;
+    }
+
+    static const char *gles_exts[] = {"GL_OES_EGL_image_external", 0};
+    hw->glsl_extensions = gles_exts;
+
+    MP_VERBOSE(hw, "using RKMPP EGL interop\n");
+
+    return 0;
+}
+
+static void mapper_unmap(struct ra_hwdec_mapper *mapper)
+{
+    struct priv *p = mapper->priv;
+
+    for (int i = 0; i < MAX_NUM_PLANES; i++) {
+        if (p->images[i])
+            p->DestroyImageKHR(eglGetCurrentDisplay(), p->images[i]);
+        p->images[i] = 0;
+    }
+}
+
+static void mapper_uninit(struct ra_hwdec_mapper *mapper)
+{
+    struct priv *p = mapper->priv;
+    GL *gl = ra_gl_get(mapper->ra);
+
+    gl->DeleteTextures(MAX_NUM_PLANES, p->gl_textures);
+    for (int i = 0; i < MAX_NUM_PLANES; i++) {
+        p->gl_textures[i] = 0;
+        ra_tex_free(mapper->ra, &p->tex[i]);
+    }
+}
+
+static int mapper_init(struct ra_hwdec_mapper *mapper)
+{
+    struct priv *p = mapper->priv;
+    GL *gl = ra_gl_get(mapper->ra);
+
+    // EGL_KHR_image_base
+    p->CreateImageKHR = (void *)eglGetProcAddress("eglCreateImageKHR");
+    p->DestroyImageKHR = (void *)eglGetProcAddress("eglDestroyImageKHR");
+    // GL_OES_EGL_image
+    p->EGLImageTargetTexture2DOES =
+        (void *)eglGetProcAddress("glEGLImageTargetTexture2DOES");
+
+    if (!p->CreateImageKHR || !p->DestroyImageKHR ||
+        !p->EGLImageTargetTexture2DOES)
+        return -1;
+
+    mapper->dst_params = mapper->src_params;
+    mapper->dst_params.imgfmt = IMGFMT_RGB0;
```

Comment:
Adding here also another comment that @wm4 made in my repo : "This makes me nervous. I'm fairly sure not all DRM hwcontext users will convert to RGB when mapping. On what does this depend?"

In the first place in introduced that because for rockchip frmaes come as an NV12 format, without dest formats, the NV12 wouldn't be converted to RGB by shaders. I have seen quite a lot of other hwdec overlays force that ro RGB0.

Also i was thinking that GL was using RGB no matter what, in which case I don't understand why this would be an issue. @wm4 : Could you please detail a bit more which use case you are thinking of that could be a problem ?

Reply:
If you'd map vaapi frames as DRM hwcontext frames, it certainly wouldn't sample it as RGB?